xds/googledirectpath/googlec2p.go

```diff
@@ -0,0 +1,222 @@
+	"google.golang.org/grpc/xds/internal/env"
+	"google.golang.org/grpc/xds/internal/version"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	_ "google.golang.org/grpc/xds" // To register xds resolvers and balancers.
+)
+
+const (
+	c2pScheme = "google-c2p"
+
+	tdURL          = "directpath-trafficdirector.googleapis.com"
+	httpReqTimeout = 10 * time.Second
+	zoneURL        = "http://metadata.google.internal/computeMetadata/v1/instance/zone"
+	ipv6URL        = "http://metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/ipv6s"
+
+	gRPCUserAgentName               = "gRPC Go"
+	clientFeatureNoOverprovisioning = "envoy.lb.does_not_support_overprovisioning"
+	ipv6CapableMetadataName         = "TRAFFICDIRECTOR_DIRECTPATH_C2P_IPV6_CAPABLE"
+
+	logPrefix = "[google-c2p-resolver]"
+)
+
+// For overriding in unittests.
+var (
+	onGCE = googlecloud.OnGCE
+
+	newClientWithConfig = func(config *bootstrap.Config) error {
+		_, err := xdsclient.NewWithConfig(config)
+		return err
+	}
+
+	logger = internalgrpclog.NewPrefixLogger(grpclog.Component("directpath"), logPrefix)
+
+	defaultNode = &v3corepb.Node{
+		Id:                   "C2P",
+		Metadata:             nil, // To be set if ipv6 is enabled.
+		Locality:             nil, // To be set to the value from metadata.
+		UserAgentName:        gRPCUserAgentName,
+		UserAgentVersionType: &v3corepb.Node_UserAgentVersion{UserAgentVersion: grpc.Version},
+		ClientFeatures:       []string{clientFeatureNoOverprovisioning},
+	}
+
+	ipv6EnabledMetadata = &structpb.Struct{
+		Fields: map[string]*structpb.Value{
+			ipv6CapableMetadataName: {
+				Kind: &structpb.Value_BoolValue{BoolValue: true},
+			},
+		},
+	}
+
+	dnsName, xdsName = "dns", "xds"
+)
+
+func init() {
+	// TODO: remove this env var check when c2p resolver is proven stable.
+	if env.C2PResolverSupport {
+		resolver.Register(&c2pResolverBuilder{})
+	}
+}
+
+type c2pResolverBuilder struct{}
+
+func (*c2pResolverBuilder) Build(t resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions) (resolver.Resolver, error) {
+	dnsB := resolver.Get(dnsName)
+	if dnsB == nil {
+		return nil, fmt.Errorf("balancer for %q is not registered", dnsName)
+	}
+	xdsB := resolver.Get(xdsName)
+	if xdsB == nil {
+		return nil, fmt.Errorf("balancer for %q is not registered", xdsName)
+	}
+	ret := &c2pResolver{
+		target: t,
+		cc:     cc,
+		opts:   opts,
+
+		dnsB: dnsB,
+		xdsB: xdsB,
+	}
+	// start does I/O. Run it in a goroutine.
+	go ret.start()
+	return ret, nil
+}
+
+func (*c2pResolverBuilder) Scheme() string {
+	return c2pScheme
+}
+
+type c2pResolver struct {
+	target resolver.Target
+	cc     resolver.ClientConn
+	opts   resolver.BuildOptions
+
+	dnsB resolver.Builder
+	xdsB resolver.Builder
+
+	mu    sync.Mutex
+	done  bool
+	child resolver.Resolver
+}
+
+func (r *c2pResolver) start() {
+	if !runDirectPath() {
+		r.startChild(r.dnsB)
+		return
+	}
```

Comment:
Is this something we could do inside `Build`?  If not, why not?

And is the only reason we avoid building the xds resolver to avoid I/O? Why is that important?

Reply:
We can block `Build()`, but the problem is we don't know how long to wait.
- The http timeout (for contact GCP metadata APIs) is 10 seconds.
- We don't get a ctx for `Build()`.